### PR TITLE
os/bluestore: deferred IO notify and locking optimization

### PR DESCRIPTION
**Pull request by majianpeng**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by majianpeng** [comment]
@ifed01 . please review.

**Comment by majianpeng** [comment]
retest this please

**Comment by markhpc** [comment]
@majianpeng can you provide a short description of what you are doing here and performance results?

Thanks kindly,
Mark

**Comment by majianpeng** [comment]
This PR mainly reduce unnecessary notify.

**Comment by majianpeng** [comment]
@markhpc  from my test(1osd), i didn't see any performance improve for 4k randread/randwrite.

**Comment by markhpc** [comment]
@majianpeng Ok, thanks!

**Comment by majianpeng** [comment]
@liewegas   . Can you review this? Thanks!

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -1682,6 +1682,7 @@ class BlueStore : public ObjectStore,
 	  auto it = q.rbegin();
 	  it++;
 	  if (it->state >= TransContext::STATE_KV_SUBMITTED) {
+	    --kv_submitted_waiters;
```
[the remark this answers is not in the document]
Comment (majianpeng) [reply src/os/bluestore/BlueStore.h]: @liewegas . Yes

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-08-09_23:09:43-rados-wip-kefu-testing-2019-08-09-2332-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-12_14:45:48-rados-wip-kefu-testing-2019-08-12-1306-distro-basic-mira/

File: src/os/bluestore/BlueStore.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/os/bluestore/BlueStore.h]: @majianpeng could you file a ticket in the tracker and connect it with this change?

[the remark this answers is not in the document]
Comment (majianpeng) [reply src/os/bluestore/BlueStore.h]: @tchaikov .  got it. i'll do it right away

[the remark this answers is not in the document]
Comment (majianpeng) [reply src/os/bluestore/BlueStore.h]: create issue https://tracker.ceph.com/issues/41216 to track this bug.

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/os/bluestore/BlueStore.h]: @majianpeng thanks! i updated the ticket. please let me know if i am wrong.

**Pull request merged**